/*
 * kongamusic (2026)
 * © Samk
 * GPL-3.0 License | Contributors: see git history
 * Do not remove or alter this notice. - Per GPL-3.0 Section 4 & Section 5
 */

package moe.kongamusic.utils

import java.math.BigInteger
import java.security.MessageDigest

fun makeTimeString(duration: Long?): String {
    if (duration == null || duration < 0) return ""

    if (duration > 1_000_000_000_000L) {
        val sdf = java.text.SimpleDateFormat("yyyy-MM-dd HH:mm:ss")
        sdf.timeZone = java.util.TimeZone.getDefault()
        return sdf.format(java.util.Date(duration))
    }

    var sec = duration / 1000
    val day = sec / 86400
    sec %= 86400
    val hour = sec / 3600
    sec %= 3600
    val minute = sec / 60
    sec %= 60

    return when {
        day > 0 -> "%dd %dh %dm %ds".format(day, hour, minute, sec)
        hour > 0 -> "%dh %dm %ds".format(hour, minute, sec)
        minute > 0 -> "%d:%02d".format(minute, sec)
        else -> "%d:%02d".format(0, sec)
    }
}

fun md5(str: String): String {
    val md = MessageDigest.getInstance("MD5")
    return BigInteger(1, md.digest(str.toByteArray())).toString(16).padStart(32, '0')
}

fun joinByBullet(vararg str: String?) =
    str
        .filterNot {
            it.isNullOrEmpty()
        }.joinToString(separator = " • ")
